core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -1197,6 +1213,78 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     KVUtils.open(newStorePath, metadata)
   }
 
+  private def createHybridStore(
+      dm: HistoryServerDiskManager,
+      appId: String,
+      attempt: AttemptInfoWrapper,
+      metadata: AppStatusStoreMetadata): KVStore = {
+
+    var retried = false
+    var hybridStore: HybridStore = null
+    while (hybridStore == null) {
+      val reader = EventLogFileReader(fs, new Path(logDir, attempt.logPath),
+        attempt.lastIndex)
+      val isCompressed = reader.compressionCodec.isDefined
+
+      // Throws an exception if the memory space is not enough
+      memoryManager.lease(appId, attempt.info.attemptId, reader.totalSize, isCompressed)
+
+      logInfo(s"Leasing disk manager space for app $appId / ${attempt.info.attemptId}...")
+      val lease = dm.lease(reader.totalSize, isCompressed)
+      val isLeaseRolledBack = new java.util.concurrent.atomic.AtomicBoolean(false)
+      var store: HybridStore = null
+      try {
+        store = new HybridStore()
+        val levelDB = KVUtils.open(lease.tmpPath, metadata)
+        store.setLevelDB(levelDB)
+        rebuildAppStore(store, reader, attempt.info.lastUpdated.getTime())
+
+        // Start the background thread to dump data to levelDB when writing to
+        // InMemoryStore is completed.
+        store.switchToLevelDB(new HybridStore.SwitchToLevelDBListener {
+          override def onSwitchToLevelDBSuccess: Unit = {
+            levelDB.close()
+            val newStorePath = lease.commit(appId, attempt.info.attemptId)
+            store.setLevelDB(KVUtils.open(newStorePath, metadata))
+            memoryManager.release(appId, attempt.info.attemptId)
+            logInfo(s"Completely switched to LevelDB for app $appId / ${attempt.info.attemptId}.")
+          }
+
+          override def onSwitchToLevelDBFail(e: Exception): Unit = {
+            levelDB.close()
+            if (!isLeaseRolledBack.getAndSet(true)) {
+              lease.rollback()
+            }
```

Comment:
I feel like this atomicBoolean variable is somewhat unnecessary. If the background thread can successfully start, the createHybridStore() will finish and return, hence no exceptions will be caught in the foreground thread.  If exceptions occurred before backgroundThread.start(), the background thread won't be started and hence the lease can only be rollbacked on the foreground thread. What do you think? @HyukjinKwon @redsanket

Reply:
Yeah LeaseRolledBack seems uneccessary